ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,612 @@
+ * chains can be checked using the {@link #checkOperationChains(List)} method. See {@link RouterServerPlaintextTest}
+ * and {@link RouterServerSSLTest} for example usage.
+ */
+class RouterServerTestFramework {
+  static final int AWAIT_TIMEOUT = 20;
+  static final int CHUNK_SIZE = 1024 * 1024;
+  private final MockClusterMap clusterMap;
+  private final MockNotificationSystem notificationSystem;
+  private final Router nonBlockingRouter;
+  private final Router coordinatorBackedRouter;
+
+  /**
+   * Instantiate a framework for testing router-server interaction. Creates a non-blocking and coordinator-backed router
+   * to interact with the passed-in {@link MockCluster}.
+   * @param routerProps All of the properties to be used when instantiating the coordinator and routers
+   * @param cluster A {@link MockCluster} that contains the servers to be used in the tests.
+   * @param notificationSystem A {@link MockNotificationSystem} that is used to determine if
+   * @throws Exception
+   */
+  RouterServerTestFramework(Properties routerProps, MockCluster cluster, MockNotificationSystem notificationSystem)
+      throws Exception {
+    this.clusterMap = cluster.getClusterMap();
+    this.notificationSystem = notificationSystem;
+
+    VerifiableProperties routerVerifiableProps = new VerifiableProperties(routerProps);
+    this.nonBlockingRouter =
+        new NonBlockingRouterFactory(routerVerifiableProps, clusterMap, notificationSystem).getRouter();
+
+    RouterConfig routerConfig = new RouterConfig(routerVerifiableProps);
+    CoordinatorBackedRouterMetrics coordinatorBackedRouterMetrics =
+        new CoordinatorBackedRouterMetrics(clusterMap.getMetricRegistry());
+    AmbryCoordinator coordinator = new HelperCoordinator(routerVerifiableProps, clusterMap);
+    this.coordinatorBackedRouter =
+        new CoordinatorBackedRouter(routerConfig, coordinatorBackedRouterMetrics, coordinator);
+  }
+
+  /**
+   * Close the instantiated routers.
+   * @throws IOException
+   */
+  void cleanup()
+      throws IOException {
+    if (nonBlockingRouter != null) {
+      nonBlockingRouter.close();
+    }
+    if (coordinatorBackedRouter != null) {
+      coordinatorBackedRouter.close();
+    }
+  }
+
+  /**
+   * Await completion of all {@link OperationChain}s in the {@code opChains} list. For each chain, check the results
+   * of each stage of the chain. Also check that the blobs put into the cluster are relatively balanced between
+   * partitions (no more than 3 times expected number of blobs per partition).
+   * @param opChains the {@link OperationChain}s to await and check.
+   * @throws Exception
+   */
+  void checkOperationChains(List<OperationChain> opChains)
+      throws Exception {
+    Map<PartitionId, Integer> partitionCount = new HashMap<>();
+    double blobsPut = 0;
+    for (OperationChain opChain : opChains) {
+      opChain.latch.await(AWAIT_TIMEOUT, TimeUnit.SECONDS);
```

Comment:
don't you need to check the return value of `await()` here and perhaps fail the test?